### PR TITLE
[PA-748] Set unset namespace labels during backup schedule

### PR DESCRIPTION
**Pull request by apimpalgaonkar**
**What this PR does / why we need it**:
* In this PR namespace labels are applied to multiple namespace and a backup schedule is created.
after backup schedule takes couple of backups now namespace label is removed from one namespace and it is verified that the namespace is not included in the next backup.
* After that the labels are applied back to the namespace and verified that the namespace is included in the next schedule backup.
  
**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PA-748

Jenkins link - https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/628/consoleFull

Aetos link -  http://aetos.pwx.purestorage.com/resultSet/testSetID/133448/testCaseID/673209

**Special notes for your reviewer**:
Added one minor change in `AddNamespaceLabel`

### CODE REVIEW

File: tests/backup/backup_restore_basic_test.go [hunk 1]
```diff
@@ -1303,3 +1303,225 @@ var _ = Describe("{BackupSyncBasicTest}", func() {
 		CleanupCloudSettingsAndClusters(backupLocationMap, credName, cloudCredUID, ctx)
 	})
 })
+
+var _ = Describe("{SetUnsetNSLabelDuringScheduleBackup}", func() {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_restore_basic_test.go:1307]: Add comment explaining what the testcase does.

File: tests/backup/backup_helper.go [hunk 2]
```diff
@@ -1445,3 +1445,122 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
...
+		return "", false, nil
+	}
+	_, err = task.DoRetryWithTimeout(backupSuccessCheck, 10*time.Minute, 30*time.Second)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func DeleteScheduleBackup(backupScheduleName, backupScheduleUID, OrgID string) error {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_helper.go:1533]: Add comment describing what this function does.

File: tests/backup/backup_helper.go [hunk 3]
```diff
@@ -1445,3 +1445,122 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
...
+	for _, namespace := range namespaces {
+		err := Inst().S.AddNamespaceLabel(namespace, labels)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return labels, nil
+}
+
+func CreateScheduleBackupUsingNamespaceLabel(scheduleName string, clusterName string, bLocation string, bLocationUID string, labelSelectors map[string]string, orgID string, preRuleName string,
```
Comment (kshithijiyer-px) [inline tests/backup/backup_helper.go:1465]: Add comment describing what this function does before the function.

File: tests/backup/backup_helper.go [hunk 4]
```diff
@@ -1445,3 +1445,122 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
 	_, err := task.DoRetryWithTimeout(backupDeletionSuccessCheck, backupDeleteTimeout, backupDeleteRetryTime)
 	return err
 }
+
+func AddLabelsToMultipleNS(number int, namespaces []string) (map[string]string, error) {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_helper.go:1449]: Add comment describing what this function does before the function.

**Review by kshithijiyer-px — CHANGES_REQUESTED** [review]

**Comment by apimpalgaonkar** [comment]
Jenkins link - 
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/839/consoleFull

File: tests/backup/backup_restore_basic_test.go [hunk 5]
```diff
@@ -1442,3 +1442,189 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+				credName = fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				err = CreateCloudCredential(provider, credName, cloudCredUID, orgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating cloud credential named %v", credName))
+				log.InfoD("Created Cloud Credentials with name - %s", credName)
+				backupLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				backupLocationMap[backupLocationUID] = backupLocationName
+				err := CreateBackupLocation(provider, backupLocationName, backupLocationUID, credName, cloudCredUID, getGlobalBucketName(provider), orgID, "")
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating backup location %v", backupLocationName))
+			}
+		})
```
Comment (skonda-px) [inline tests/backup/backup_restore_basic_test.go:1538]: Why cloud creds and backuplocation add twice?

File: tests/backup/backup_restore_basic_test.go [hunk 6]
```diff
@@ -1442,3 +1442,189 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+			periodicSchedulePolicyUid, err = Inst().Backup.GetSchedulePolicyUid(orgID, ctx, periodicSchedulePolicyName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching uid of periodic schedule policy named [%s]", periodicSchedulePolicyName))
+		})
+		Step("Creating a schedule backup", func() {
+			log.InfoD("Creating a schedule backup")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Unable to fetch px-central-admin ctx")
+			scheduleName = fmt.Sprintf("%s-schedule-%v", BackupNamePrefix, time.Now().Unix())
+			err = CreateScheduleBackupWithNamespaceLabel(scheduleName, SourceClusterName, backupLocationName, backupLocationUID,
+				labelSelectors, orgID, "", "", "", "", nsLabelString, periodicSchedulePolicyName, periodicSchedulePolicyUid, ctx)
```
Comment (skonda-px) [inline tests/backup/backup_restore_basic_test.go:1571]: Remove labelSelectors if there are label Selectors

**Review by skonda-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_restore_basic_test.go [hunk 7]
```diff
@@ -1442,3 +1442,189 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of schedule backup with schedule name [%s]", scheduleName))
+			firstScheduleBackupName, err := GetFirstScheduleBackupName(ctx, scheduleName, orgID)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching the name of the first schedule backup [%s]", firstScheduleBackupName))
+			err = NamespaceLabelBackupSuccessCheck(firstScheduleBackupName, ctx, bkpNamespaces, nsLabelString)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if the labeled namespace [%v] is backed up and checks for labels [%s] applied to backup [%s]", bkpNamespaces, nsLabelString, firstScheduleBackupName))
+			log.InfoD("Waiting for 15 minutes for the next schedule backup to be triggered")
+			time.Sleep(2 * time.Minute)
```
Comment (vpinisetti-px) [inline tests/backup/backup_restore_basic_test.go:1578]: As per log message its 15 minutes. But in sleep func it updated as 2 minutes. So with 2 minutes also its working.

**Review by vpinisetti-px — CHANGES_REQUESTED** [review]

**Comment by apimpalgaonkar** [comment]
Jenkins link - 
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1014/console

**Review by vpinisetti-px — APPROVED** [review]
LGTM

File: tests/backup/backup_restore_basic_test.go [hunk 8]
```diff
@@ -1442,3 +1442,170 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+			err = CreateScheduleBackupWithNamespaceLabel(scheduleName, SourceClusterName, backupLocationName, backupLocationUID,
+				nil, orgID, "", "", "", "", nsLabelString, periodicSchedulePolicyName, periodicSchedulePolicyUid, ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of schedule backup with schedule name [%s]", scheduleName))
+			firstScheduleBackupName, err := GetFirstScheduleBackupName(ctx, scheduleName, orgID)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching the name of the first schedule backup [%s]", firstScheduleBackupName))
+			err = NamespaceLabelBackupSuccessCheck(firstScheduleBackupName, ctx, bkpNamespaces, nsLabelString)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if the labeled namespace [%v] is backed up and checks for labels [%s] applied to backup [%s]", bkpNamespaces, nsLabelString, firstScheduleBackupName))
+			log.InfoD("Waiting for 15 minutes for the next schedule backup to be triggered")
+			time.Sleep(15 * time.Minute)
+			secondScheduleBackupName, err := GetOrdinalScheduleBackupName(ctx, scheduleName, 2, orgID)
```
Comment (skonda-px) [inline tests/backup/backup_restore_basic_test.go:1560]: is there any difference between GetNextScheduleBackupName and these 3 lines of code.

Comment (apimpalgaonkar) [reply tests/backup/backup_restore_basic_test.go:1560]: GetNextScheduleBackupName this is executed in the middle of schedule where we are not sure at which point the schedule backup is running, where it first fetches all backups and then get the next backup, but in the above case we need two subsequent schedule backups so I used this one.

File: tests/backup/backup_helper.go [hunk 9]
```diff
@@ -2188,3 +2188,31 @@ func AreSlicesEqual(slice1, slice2 interface{}) bool {
 	}
 	return true
 }
+
```
Comment (kshithijiyer-px) [inline tests/backup/backup_helper.go:2191]: Add a doc string for the func

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

**Pull request merged**